tensorboard/webapp/plugins/npmi/data_source/npmi_data_source.ts

```diff
@@ -40,6 +54,61 @@ export class NpmiHttpServerDataSource implements NpmiDataSource {
       this.fetchAnnotations(),
       this.fetchMetrics(),
       this.fetchValues()
+    ).pipe(
+      map(([annotations, metrics, values]) => {
+        let annotationData: AnnotationDataListing = {};
+        for (let run of Object.keys(annotations)) {
+          for (let annotationIndex in annotations[run]) {
+            let annotation = annotations[run][annotationIndex];
+            let dataElements: ValueData[] = [];
+            for (let metricIndex in metrics[run]) {
+              let metric = metrics[run][metricIndex];
+              let dataElement = dataElements.find(
+                (element) =>
+                  element.metric === metric_type.stripMetricString(metric)
+              );
```

Comment:
I worry a bit about the inefficiency here; this can be an expensive operation. Can we instead use `Map`?

```ts
for (let annotationIndex in annotations[run]) {
  let annotation = annotations[run][annotationIndex];
  const metricToDataElements = new Map<string, ValueData>();

  for (let metricIndex in metrics[run]) {
    const metric = metrics[run][metricIndex];
    const metricString = metric_type.stripMetricString(metric);

    let dataElement = metricToDataElements.get(metricString);

    if (!dataElement) {
      dataElement = {
        nPMIValue: null,
        countValue: null,
        annotation: annotation,
        metric: metric_type.stripMetricString(metric),
        run: run,
      };
      metricToDataElements.set(metricString, dataElement);
    }

    if (metric_type.metricIsMetricCount(metric)) {
      dataElement.countValue = values[run][annotationIndex][metricIndex];
    } else if (metric_type.metricIsNpmi(metric)) {
      dataElement.nPMIValue = values[run][annotationIndex][metricIndex];
    }
  }

  annotationData[annotation] = [
    ...annotationData[annotation],
    ...metricToDataElements.values(),
  ];
}
```

Reply:
Thank you for the hint! 👍